bin/dmarc_view_reports

```diff
@@ -79,32 +80,46 @@ sub print_header {
     return;
 };
 
-sub print_geoip {
+sub get_geoip_details {
     my $ip = shift;
-    return if ! $geoip_opt;
+
+    return if ! defined $geoip_opt;
+    $geoip_opt ||= 'country_name';
+
     $gip ||= get_geoip_db();
     return if ! $gip;
 
-    my $r = $gip->record_by_addr($ip) or return '';
-    my $country_code = $r->country_code();
-    my $country_name = $r->country_name();
-    my $continent = $r->continent_code();
-    my $city = $r->city;
-    print $city ?  " $city," : ' ';
-    print " $country_code, $continent";
-    return;
-}
+    if ($ip =~ /^\d+\.\d+\.\d+\.\d+$/) {
+        $ip = '::ffff:'.$ip;
+    }
 
-sub get_dns_hostname {
-    my $ip = shift;
-    return if ! $dns_opt;
+    my $r = $gip->record_by_addr_v6($ip) or return '';
+
+    my @result;
+    my @fields = split(',', $geoip_opt);
+    my @allowed = qw(
+	country_code
+	country_code3
+	country_name
+	region
+	region_name
+	city
+	postal_code
+	latitude
+	longitude
+	time_zone
+	area_code
+	continent_code
+	metro_code
+    );
 
-    my @answers = $report->has_dns_rr('PTR', $ip);
-    return '' if 0 == scalar @answers;
-    return $answers[0] if scalar @answers >= 1;
-    print Dumper(\@answers);
-    return;
-};
+    foreach my $f (@fields) {
+       return if ! grep {$_ eq $f} @allowed;
```

Comment:
this should be `next` instead of `return`. Return exits the subroutine instead of the foreach loop.